### DISCUSSION TITLE
Derive with a parameter?

### DISCUSSION THREAD

**Discussion by rawberg** [pmndrs/valtio#770] [Q&A] [answered]
Is there a pattern to pass a parameter into derive function? this would allow the user to choose an option that would then filter store data. prefer not to save the user choice in the store as it's not global state and there are multiple different filtering options.

something like:
```
const getCollectionBy = derive({
   title: (get, val) => _getCollectionByTitle(val, get(collectionState).tree),
})
```

**Comment by dai-shi** [comment]
hmm,  the derive function is declarative, so you don't/can't control when it's invoked.
I wouldn't use `derive` for this case.
```js
const getCollectionBy = proxy({
  title: null,
  setTitle: (val) => {
    getCollectionBy.title = _getCollectionByTitle(val, collectionState.tree)
  },
})
```

**Reply by rawberg** [reply]
thank you! that solution does work. however if that's in the global store next to the `collectionState` proxy that means if any component calls `getCollectionBy.setTitle("myfilter")` then all the components that are using the value from `getCollectionBy.title` will be updated to use that data. is there another approach that would allow a parameter to be passed without needing to set it anywhere so different components can be subscribed to different filters of `collectionState` in this example?

**Reply by rawberg** [reply]
does it make sense to have `getCollectionBy` be a function that returns a proxy, then each caller effectively has their own copy of it?

```javascript
const getCollectionBy = () => proxy({
  title: null,
  setTitle: (val) => getCollectionBy.title = _getCollectionByTitle(val, collectionState.tree)
})
```

**Reply by rawberg** [reply]
returning a new proxy for each component, won't auto re-render via `useSnapshot`

**Reply by dai-shi** [reply, marked as answer]
> so different components can be subscribed to different filters

Hmm? If it's not global, you can use `useState` or `useRef+proxy` for each component.
But, it sounds like you don't even need local state for this. Just derive (not valtio one) from what you already have.

I'm not sure the exact use case, but `derive` from valtio shouldn't be a tool for this case.

**Reply by rawberg** [reply]
Thank for your help, you're right. I solved it by calling my filter function from within the react component and passing in state from `useSnapshot`

```
const snap = useSnapshot(myModule.myProxy);
return (
  <ul>
    <Collection collection={myModule.getCollectionByTitle(props.options.collectionName, snap.tree)}/>
  </ul>
```
